File: objc/main/test/objc/TestTreeConstruction.m
#import "TestTreeConstruction.h"

@implementation TestTreeConstruction

- (void) setup {
  lexerClassName = @"org.stringtemplate.v4.compiler.STLexer";
  parserClassName = @"org.stringtemplate.v4.compiler.STParser";
}

- (void) test_template1 {
  RuleReturnScope *rstruct = (RuleReturnScope *)[self execParser:@"template" arg1:@"<[]>" arg2:16];
  NSObject *actual = [((Tree *)[rstruct tree]) toStringTree];
  NSObject *expecting = @"(EXPR [)";
  [self assertEquals:@"testing rule template" arg1:expecting arg2:actual];
}

- (void) test_template2 {
  RuleReturnScope *rstruct = (RuleReturnScope *)[self execParser:@"template" arg1:@"<[a,b]>" arg2:17];
  NSObject *actual = [((Tree *)[rstruct tree]) toStringTree];
  NSObject *expecting = @"(EXPR ([ a b))";
  [self assertEquals:@"testing rule template" arg1:expecting arg2:actual];
}

@end

File: java/test/org/stringtemplate/v4/test/TestTreeConstruction.java
package org.stringtemplate.v4.test;

import org.antlr.runtime.RuleReturnScope;
import org.antlr.runtime.tree.Tree;

import org.junit.*;

import static org.junit.Assert.assertEquals;

public class TestTreeConstruction extends gUnitBase {
	@Before public void setup() {
	    lexerClassName = "org.stringtemplate.v4.compiler.STLexer";
	    parserClassName = "org.stringtemplate.v4.compiler.STParser";
	}
	@Test public void test_template1() throws Exception {
		// gunit test on line 16
		RuleReturnScope rstruct = (RuleReturnScope)execParser("template", "<[]>", 16);
		Object actual = ((Tree)rstruct.getTree()).toStringTree();
		Object expecting = "(EXPR [)";
		assertEquals("testing rule template", expecting, actual);
	}

	@Test public void test_template2() throws Exception {
		// gunit test on line 17
		RuleReturnScope rstruct = (RuleReturnScope)execParser("template", "<[a,b]>", 17);
		Object actual = ((Tree)rstruct.getTree()).toStringTree();
		Object expecting = "(EXPR ([ a b))";
		assertEquals("testing rule template", expecting, actual);
	}
}
